### PR TITLE
reuse solver, fixes #541

### LINKED ISSUE #541: Reuse of amici::Solver
The original plan was reusing amici::Solver objects for multiple simulations. However, this sometimes causes problems, at least in cases of previous integration failures. Did not dig too deep, but for example setting tolerances for backward integration after previous failures causes CVODES errors (invalid `which`), probably related to improper checking of `cv_adjMallocDone` or similar. 

If we want to have Solver reusable, this needs some cleanup/fixing. For transparent (temporary) workaround, we could simply clone Solver in `runAmiciSimulation`.

### PR DESCRIPTION
**Pull request by FFroehlich**

### CODE REVIEW

File: include/amici/solver.h [hunk 1]
```diff
@@ -490,32 +399,45 @@ class Solver {
...
      * @param maxsteps maximum number of solver steps (non-negative number)
      */
-    void setMaxSteps(int maxsteps);
+    void setMaxSteps(const long int maxsteps);
+
+    /**
+     * @brief sets the maximum number of solver steps for the forward problem
+     * @param maxsteps maximum number of solver steps (non-negative number)
+     */
+    void setMaxSteps(const int maxsteps);
```
Comment (dweindl) [inline include/amici/solver.h]: Not really necessary, since int would automatically be widened to long int.

File: include/amici/solver.h [hunk 2]
```diff
@@ -490,32 +399,45 @@ class Solver {
...
+     * @param maxsteps maximum number of solver steps (non-negative number)
+     */
+    void setMaxStepsBackwardProblem(const int maxsteps);
+
     /**
      * @brief sets the maximum number of solver steps for the backward problem
      * @param maxsteps maximum number of solver steps (non-negative number)
      */
-    void setMaxStepsBackwardProblem(int maxsteps);
+    void setMaxStepsBackwardProblem(const long int maxsteps);
```
Comment (dweindl) [inline include/amici/solver.h]: Not really necessary, since int would automatically be widened to long int.

File: include/amici/solver.h [hunk 3]
```diff
@@ -610,115 +662,173 @@ class Solver {
...
+     * @param itask task identifier, can be CV_NORMAL or CV_ONE_STEP
+     * @param ncheckPtr pointer to a number that counts the internal
+     * checkpoints
+     * @return status flag indicating success of execution
+     */
+    virtual int solveF(const realtype tout, const int itask,
+                       int *ncheckPtr) const = 0;
+    
+    /**
+     * @brief reInitPostProcessF
```
Comment (dweindl) [inline include/amici/solver.h]: More descriptive?

File: include/amici/solver.h [hunk 4]
```diff
@@ -610,115 +662,173 @@ class Solver {
...
+     * @param tnext next timepoint (defines integration direction)
+     */
+    virtual void reInitPostProcessF(const realtype tnext) const = 0;
+
+    /**
+     * @brief reInitPostProcessB
```
Comment (dweindl) [inline include/amici/solver.h]: More descriptive?

File: include/amici/solver.h [hunk 5]
```diff
@@ -610,115 +662,173 @@ class Solver {
...
+     * @param tnext next timepoint (defines integration direction)
+     */
+    virtual void reInitPostProcessB(const realtype tnext) const = 0;
+
+    /**
+     * @brief sets sx to the state sensitivity at the current timepoint
```
Comment (dweindl) [inline include/amici/solver.h]: Gets? (Also the following ones)

File: include/amici/solver.h [hunk 6]
```diff
@@ -1163,11 +1372,11 @@ class Solver {
     StateOrdering ordering = StateOrdering::AMD;
 
     /** maximum number of allowed Newton steps for steady state computation */
-    int newton_maxsteps = 0;
+    long int newton_maxsteps = 0;
```
Comment (dweindl) [inline include/amici/solver.h]: Is this explicitly long int in sundials, or something like sun_index_type? The latter would be more flexible.

File: include/amici/steadystateproblem.h [hunk 7]
```diff
@@ -97,32 +97,39 @@ class SteadystateProblem {
...
-        this->t = t;
-        this->x = x;
-        this->sx = sx;
-    }
+    void writeSolution(realtype *t, AmiVector &x, AmiVectorArray &sx) const;
+    
 
   private:
-    realtype *t;
+    realtype t;
```
Comment (dweindl) [inline include/amici/steadystateproblem.h]: doxygen?

File: include/amici/vector.h [hunk 8]
```diff
@@ -3,320 +3,290 @@
...
-    /** resets the Vector by filling with zero values
-      */
-    void reset() {
-        set(0.0);
-    }
+    /**
+     * @brief returns the length of the vector
+     * @return length
+     */
+    int getLength() const;
```
Comment (dweindl) [inline include/amici/vector.h]: Stray comment?

File: src/solver.cpp [hunk 9]
```diff
@@ -789,47 +804,194 @@ InternalSensitivityMethod Solver::getInternalSensitivityMethod() const {
...
+        getSensDky(t, 0);
+        return sx;
+    }
+}
+
+const AmiVector& Solver::getAdjointState(const int which, const realtype t) const {
+    if (t == this->t) {
+        getB(which);
+        return xB;
+    } else {
```
Comment (dweindl) [inline src/solver.cpp]: Superfluous `else` after `return`, also below

File: src/solver_cvodes.cpp [hunk 10]
```diff
@@ -346,237 +395,269 @@ void CVodeSolver::reInitPostProcess(void *ami_mem, realtype *t,
...
+    int status = CVodeGetB(solverMemory.get(), which, &tDummy, xB.getNVector());
+    if (status != CV_SUCCESS)
+        throw CvodeException(status, "CVodeGetB");
+}
+
+int CVodeSolver::solve(const realtype tout, const int itask) const {
+    if (forceReInitPostProcessF)
+        reInitPostProcessF(tout);
+    int status = CVode(solverMemory.get(), tout, x.getNVector(), &t, itask);
+    if (status < 0)
```
Comment (dweindl) [inline src/solver_cvodes.cpp]: `> 0` is okay?

File: src/solver_cvodes.cpp [hunk 11]
```diff
@@ -346,237 +395,269 @@ void CVodeSolver::reInitPostProcess(void *ami_mem, realtype *t,
...
-    }
-
-    solverWasCalled = true;
+int CVodeSolver::solveF(const realtype tout, const int itask,
+                        int *ncheckPtr) const {
+    if (forceReInitPostProcessF)
+        reInitPostProcessF(tout);
+    int status =
+        CVodeF(solverMemory.get(), tout, x.getNVector(), &t, itask, ncheckPtr);
+    if (status < 0)
```
Comment (dweindl) [inline src/solver_cvodes.cpp]: `> 0` is okay?

File: include/amici/solver.h (hunk 6 above)
Comment (FFroehlich) [reply include/amici/solver.h]: explicitely `long int`

**Pull request merged**